lib/senders/email.js

```diff
@@ -253,97 +272,306 @@ module.exports = function (log, config) {
+                            consumedPercentages[regex] = percentage
+                          }
+                        } else {
+                          globalConsumedPercentage += percentage
+                        }
+                      }
+
+                      if (isMatched) {
+                        return
+                      }
+
+                      isMatched = result
+
+                      if (isMatched) {
+                        upsertServicesMap(services, SERVICES.external[key], emailAddress, {
+                          mailer: this.emailService,
+                          emailService: 'fxa-email-service',
+                          emailSender: key
+                        })
+                      }
+                    })
+                }, promise)
+              }
+            })
+            .then(() => {
+              if (isMatched) {
+                return services
+              }
+
+              if (config.emailService.forcedEmailAddresses.test(emailAddress)) {
+                return upsertServicesMap(services, SERVICES.external.ses, emailAddress, {
+                  mailer: this.emailService,
+                  emailService: 'fxa-email-service',
+                  emailSender: 'ses'
+                })
+              }
+
+              return upsertServicesMap(services, SERVICES.internal, emailAddress, {
+                mailer: this.mailer,
+                emailService: 'fxa-auth-server',
+                emailSender: 'ses'
+              })
+            })
+        }, P.resolve(new Map()))
+      })
+      .then(services => Array.from(services.values()))
+
+    function isLiveConfigMatch (liveConfig, emailAddress, globalConsumedPercentage, consumedPercentages) {
+      return new P(resolve => {
+        let { percentage } = liveConfig
+        const { regex } = liveConfig
+
+        if (regex) {
+          if (consumedPercentages[regex]) {
+            percentage += consumedPercentages[regex]
+          }
+        } else {
+          percentage += globalConsumedPercentage
+        }
+
+        if (percentage >= 0 && percentage < 100) {
+          let sum = 0
+
+          for (const character in emailAddress) {
+            sum += emailAddress.charCodeAt(character)
+          }
+
+          if (sum % 100 >= percentage) {
+            resolve(false)
+            return
+          }
         }
-        log.info(
-          {
-            op: 'mailer.send.1',
-            status: status && status.message,
-            id: status && status.messageId,
-            to: emailConfig && emailConfig.to
+
+        if (regex) {
+          if (! safeRegex(regex)) {
+            resolve(false)
+            return
           }
-        )
 
-        emailUtils.logEmailEventSent(log, message)
+          // Execute the regex inside a sandbox and kill it if it takes > 100 ms
+          const sandbox = new Sandbox({ timeout: 100 })
+          sandbox.run(`new RegExp("${regex}").test("${emailAddress}")`, output => {
```

Comment:
emailAddress is validated before it reaches here, right? I'm wondering if someone could shove some garbage in there that causes us to run an arbitrary script in the sandbox.

Reply:
It's come from the db so was validated during account creation or when setting up secondary email. I'm pretty comfortable with that but also happy to add more validation if it's a concern.